homeassistant/components/picnic/coordinator.py

```diff
@@ -0,0 +1,142 @@
+"""Coordinator to fetch data from the Picnic API."""
+import copy
+from datetime import timedelta
+import logging
+
+import async_timeout
+from python_picnic_api import PicnicAPI
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import (
+    ADDRESS,
+    SENSOR_CART_ITEMS_COUNT,
+    SENSOR_CART_TOTAL_PRICE,
+    SENSOR_COMPLETED_DELIVERIES,
+    SENSOR_LAST_ORDER_DELIVERY_TIME,
+    SENSOR_LAST_ORDER_ETA_END,
+    SENSOR_LAST_ORDER_ETA_START,
+    SENSOR_LAST_ORDER_SLOT_END,
+    SENSOR_LAST_ORDER_SLOT_START,
+    SENSOR_LAST_ORDER_STATUS,
+    SENSOR_LAST_ORDER_TOTAL_PRICE,
+    SENSOR_SELECTED_SLOT_END,
+    SENSOR_SELECTED_SLOT_MAX_ODER_TIME,
+    SENSOR_SELECTED_SLOT_MIN_ORDER_VALUE,
+    SENSOR_SELECTED_SLOT_START,
+    SENSOR_TOTAL_ORDERS,
+)
+
+
+class PicnicUpdateCoordinator(DataUpdateCoordinator):
+    """The coordinator to fetch data from the Picnic API at a set interval."""
+
+    def __init__(self, hass: HomeAssistant, picnic_api_client: PicnicAPI):
+        """Initialize the coordinator with the given Picnic API client."""
+        self.picnic_api_client = picnic_api_client
+
+        logger = logging.getLogger(__name__)
+        super().__init__(
+            hass,
+            logger,
+            name="Picnic coordinator",
+            update_interval=timedelta(minutes=30),
+            update_method=self.async_update_data,
+        )
+
+    async def async_update_data(self):
+        """Fetch data from API endpoint."""
+        try:
+            # Note: asyncio.TimeoutError and aiohttp.ClientError are already
+            # handled by the data update coordinator.
+            async with async_timeout.timeout(10):
+                return await self.hass.async_add_executor_job(self.fetch_data)
+        except Exception as err:  # pylint: disable=broad-except
```

Comment:
We shouldn't await a broad exception. What exceptions can your PyPI lib raise ?

Reply:
I've changed it to more specific Exceptions, the lib returns a ValueError when the response is not proper JSON and a  PicnicAuthError when authentication failed.